### PR TITLE
[JIT] Traceable detach

### PR DESCRIPTION
**Pull request by jamesr66a**
This makes it so `detach` and `detach_` are traceable and also adds a pass to erase them before ONNX export

### CODE REVIEW

File: torch/csrc/jit/passes/erase_detach.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "torch/csrc/jit/ir.h"
+
+namespace torch { namespace jit {
+
+// Erase detach. This is necessary for export to ONNX, since ONNX
+// does not support training. This is safe because, again,
+// ONNX does not suppor ttracing.
+TORCH_API void EraseDetach(const std::shared_ptr<Graph>& graph);
```
Comment (zdevito) [inline torch/csrc/jit/passes/erase_detach.h]: You don't need an entire pass to do this. Here is how it is done for a similar op, clone in `symbolics.py`:

```
# ignore clone operators that are inserted by PyTorch autograd
def clone(g, input):
    return input
```

**Review by zdevito — COMMENTED** [review]
The removal pass is unnecessary, but this otherwise looks good.

**Review by zdevito — APPROVED** [review]

File: test/test_jit.py [hunk 2]
```diff
@@ -1396,6 +1396,37 @@ def constant_prop():
         self.run_pass('constant_propagation', constant_prop.graph)
         self.assertExpected(canonical(constant_prop.graph))
 
+    def test_trace_detach(self):
+        def foo(x, w):
+            return torch.matmul(x, w).detach()
+
+        traced = torch.jit.trace(torch.rand(3, 4), torch.rand(4, 5))(foo)
+
+        self.assertExpected(str(traced.graph))
```
Comment (apaszke) [inline test/test_jit.py]: `assertExpectedGraph`

File: test/test_jit.py [hunk 3]
```diff
@@ -1396,6 +1396,37 @@ def constant_prop():
...
+        x, w = torch.rand(3, 4), torch.rand(4, 5)
+        self.assertEqual(foo(x, w), traced(x, w))
```
Comment (apaszke) [inline test/test_jit.py]: Are you sure `assertEqual` checks the `requires_grad` status? I'm pretty sure it only cares about the data, so you need to do it yourself

**Review by apaszke — DISMISSED** [review]

**Comment by apaszke** [comment]
NB: this also needs to be rebased on top of master, because it uses the legacy tracing API in the test
